File: infrastructures/ddd/csharp/src/Domain/Events/BaseDomainEvent.cs
using System;

namespace OddlyDdd.Domain.Events
{
    /// <summary>
    /// Base class for all domain events.
    /// Domain events MUST:
    /// - Suffix with "Event"
    /// - Follow pattern: {Object}{Action}Event (e.g., ExampleCreatedEvent)
    /// - Be immutable (use readonly fields or init-only properties)
    /// - Live in /Domain/Events/
    /// - Include timestamp and correlationId
    /// </summary>
    public abstract class BaseDomainEvent
    {
        /// <summary>
        /// Unique identifier for this event instance
        /// </summary>
        public string EventId { get; init; }

        /// <summary>
        /// Timestamp when the event occurred
        /// </summary>
        public DateTime Timestamp { get; init; }

        /// <summary>
        /// Correlation ID for tracking requests across services
        /// </summary>
        public string CorrelationId { get; init; }

        protected BaseDomainEvent(string p_correlationId)
        {
            EventId = Guid.NewGuid().ToString();
            Timestamp = DateTime.UtcNow;
            CorrelationId = p_correlationId;
        }
    }
}

File: infrastructures/ddd/java/src/main/java/com/oddly/ddd/domain/events/infra/BaseDomainEvent.java
package com.oddly.ddd.domain.events.infra;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * Base class for all domain events.
 * Domain events MUST:
 * - Have Event suffix
 * - Follow pattern: {Object}{Action}Event (e.g., ExampleCreatedEvent)
 * - Be immutable (all fields final)
 * - Include timestamp and correlationId
 * - Live in /domain/events/
 */
public abstract class BaseDomainEvent {
    private final String m_eventId;
    private final LocalDateTime m_timestamp;
    private final String m_correlationId;

    protected BaseDomainEvent(String p_correlationId) {
        this.m_eventId = UUID.randomUUID().toString();
        this.m_timestamp = LocalDateTime.now();
        this.m_correlationId = p_correlationId;
    }

    public String getEventId() {
        return m_eventId;
    }

    public LocalDateTime getTimestamp() {
        return m_timestamp;
    }

    public String getCorrelationId() {
        return m_correlationId;
    }

    /**
     * Gets the event type name.
     * Used for routing and logging.
     */
    public abstract String getEventType();
}
